swarm/network/stream/messages.go

```diff
@@ -96,26 +104,29 @@ func (p *Peer) handleSubscribeMsg(ctx context.Context, req *SubscribeMsg) (err e
 			// exchange between peers over p2p. Instead, error will be returned
 			// only if there is one from sending subscribe error message.
 			err = p.Send(context.TODO(), SubscribeErrorMsg{
-				Error: err.Error(),
+				Server: true,
+				Stream: req.Stream,
+				Error:  err.Error(),
 			})
 		}
 	}()
 
 	log.Debug("received subscription", "from", p.streamer.addr, "peer", p.ID(), "stream", req.Stream, "history", req.History)
 
-	f, err := p.streamer.GetServerFunc(req.Stream.Name)
+	os, err := p.getServer(req.Stream)
 	if err != nil {
 		return err
 	}
 
-	s, err := f(p, req.Stream.Key, req.Stream.Live)
-	if err != nil {
-		return err
-	}
-	os, err := p.setServer(req.Stream, s, req.Priority)
-	if err != nil {
-		return err
+	/*if os.est && os.stream.String() == req.Stream.String() {
+		log.Debug("stream already exists between peers, dont panic", "peer", p.ID(), "other", p.streamer.addr)
+	}*/
+
+	// todo: i think this ^^ option is better but need to see how this affects current batches
+	if os.est {
+		return fmt.Errorf("peer subscription already established, stream %s, peer %s, other %s", req.Stream, p.ID(), p.streamer.addr)
 	}
+	os.est = true
```

Comment:
I think this setting to `true` is redundant, the one on line 150 is sufficient. I tend to want to keep the latter because there is some logic which is being checked on line 144, although I am not so sure why and if that check is needed.